crates/holochain/src/core/workflow/app_validation_workflow/tests.rs

```diff
@@ -0,0 +1,491 @@
+    core::state::dht_op_integration::IntegratedDhtOpsValue,
+    core::state::validation_db::ValidationLimboValue,
+    core::{
+        state::element_buf::ElementBuf,
+        workflow::incoming_dht_ops_workflow::IncomingDhtOpsWorkspace,
+    },
+    test_utils::host_fn_api::*,
+    test_utils::setup_app,
+};
+use ::fixt::prelude::*;
+use fallible_iterator::FallibleIterator;
+use holo_hash::{AnyDhtHash, DhtOpHash, EntryHash, HeaderHash};
+use holochain_serialized_bytes::{SerializedBytes, SerializedBytesError};
+use holochain_state::{env::EnvironmentWrite, fresh_reader_test};
+use holochain_types::{
+    app::InstalledCell, cell::CellId, dht_op::DhtOpLight, dna::DnaDef, dna::DnaFile, fixt::*,
+    test_utils::fake_agent_pubkey_1, test_utils::fake_agent_pubkey_2, validate::ValidationStatus,
+    Entry,
+};
+use holochain_wasm_test_utils::TestWasm;
+use holochain_zome_types::{element::Element, zome::ZomeName, ExternInput, Header};
+use std::{
+    convert::{TryFrom, TryInto},
+    time::Duration,
+};
+use tracing::*;
+
+/// Wait for a maximum of 10 seconds
+/// for validation to run. (100 * 100ms)
+const NUM_ATTEMPTS: usize = 100;
+
+#[tokio::test(threaded_scheduler)]
+async fn app_validation_workflow_test() {
+    observability::test_run().ok();
+
+    let dna_file = DnaFile::new(
+        DnaDef {
+            name: "app_validation_workflow_test".to_string(),
+            uuid: "ba1d046d-ce29-4778-914b-47e6010d2faf".to_string(),
+            properties: SerializedBytes::try_from(()).unwrap(),
+            zomes: vec![TestWasm::Validate.into(), TestWasm::ValidateLink.into()].into(),
+        },
+        vec![TestWasm::Validate.into(), TestWasm::ValidateLink.into()],
+    )
+    .await
+    .unwrap();
+
+    let alice_agent_id = fake_agent_pubkey_1();
+    let alice_cell_id = CellId::new(dna_file.dna_hash().to_owned(), alice_agent_id.clone());
+    let alice_installed_cell = InstalledCell::new(alice_cell_id.clone(), "alice_handle".into());
+
+    let bob_agent_id = fake_agent_pubkey_2();
+    let bob_cell_id = CellId::new(dna_file.dna_hash().to_owned(), bob_agent_id.clone());
+    let bob_installed_cell = InstalledCell::new(bob_cell_id.clone(), "bob_handle".into());
+
+    let mut dna_store = MockDnaStore::new();
+
+    dna_store.expect_get().return_const(Some(dna_file.clone()));
+    dna_store.expect_add_dnas::<Vec<_>>().return_const(());
+    dna_store.expect_add_entry_defs::<Vec<_>>().return_const(());
+    dna_store.expect_get_entry_def().return_const(None);
+
+    let (_tmpdir, _app_api, handle) = setup_app(
+        vec![(
+            "test_app",
+            vec![(alice_installed_cell, None), (bob_installed_cell, None)],
+        )],
+        dna_store,
+    )
+    .await;
+
+    run_test(alice_cell_id, bob_cell_id, handle.clone(), &dna_file).await;
+
+    let shutdown = handle.take_shutdown_handle().await.unwrap();
+    handle.shutdown().await;
+    shutdown.await.unwrap();
+}
+
+async fn run_test(
```

Comment:
Could you just explain the gist of what this mega test is doing at each step? Also, maybe it would flow a bit better if the test weren't interrupted with function definitions and those got moved outside the test function?